samples/cellular/modem_shell/src/sock/sock.c

```diff
@@ -1305,6 +1305,263 @@ int sock_rai(int socket_id, bool rai_last, bool rai_no_data,
 	return 0;
 }
 
+int sock_setopt(int socket_id, int sock_level, int sock_opt_id, char *sock_opt_value)
+{
+	struct sock_info *socket_info = get_socket_info_by_id(socket_id);
+	int err;
+	char *end;
+
+	if (socket_info == NULL) {
+		return -EINVAL;
+	}
+
+	if (sock_opt_id == 0) {
+		mosh_error("Socket option is mandatory.");
+		return -EINVAL;
+	}
+
+	if (sock_opt_value == NULL &&
+	    !(sock_level == SOL_TLS && sock_opt_id == TLS_HOSTNAME) &&
+	    !(sock_level == SOL_TLS && sock_opt_id == TLS_SESSION_CACHE_PURGE) &&
+	    !(sock_level == SOL_TLS && sock_opt_id == TLS_DTLS_CONN_SAVE) &&
+	    !(sock_level == SOL_TLS && sock_opt_id == TLS_DTLS_CONN_LOAD)) {
+		mosh_error("Socket option value is mandatory.");
+		return -EINVAL;
+	}
+
+	switch (sock_level) {
+	case SOL_TLS:
+		switch (sock_opt_id) {
+		case TLS_SEC_TAG_LIST:
+		case TLS_CIPHERSUITE_LIST:
+			/* Value is a comma separated list of integers or hexadecimals, for example
+			 * "42,43,44" or "0xC024,0xC00A,0xC023".
+			 */
+			int value_list[20];
+			unsigned int value_count = 0;
+
+			for (int i = 0; i < strlen(sock_opt_value); i++) {
+				if (sock_opt_value[i] == ',') {
+					sock_opt_value[i] = ' ';
+				}
+			}
+
+			while (value_count < ARRAY_SIZE(value_list)) {
+				int value = strtoul(sock_opt_value, &end, 0);
+
+				if (end == sock_opt_value) {
+					break;
+				}
+				value_list[value_count++] = value;
+				sock_opt_value = end;
+			}
+
+			err = setsockopt(socket_info->fd, sock_level, sock_opt_id, value_list,
+					 value_count * sizeof(int));
+			goto exit;
+
+		case TLS_HOSTNAME:
+			/* Value can be empty to clear hostname. */
+			size_t hostname_len;
+
+			if (sock_opt_value == NULL) {
+				hostname_len = 0;
+			} else {
+				hostname_len = strlen(sock_opt_value);
+			}
+			err = setsockopt(socket_info->fd, sock_level, sock_opt_id, sock_opt_value,
+					 hostname_len);
+			goto exit;
+
+		case TLS_SESSION_CACHE_PURGE:
+		case TLS_DTLS_CONN_SAVE:
+		case TLS_DTLS_CONN_LOAD:
+			/* No value used for these options. */
+			err = setsockopt(socket_info->fd, sock_level, sock_opt_id, NULL, 0);
+			goto exit;
+
+		default:
+			break;
+		}
+		break; /* SOL_TLS */
+
+	case SOL_SOCKET:
+		switch (sock_opt_id) {
+		case SO_RCVTIMEO:
+		case SO_SNDTIMEO:
+			/* Value is in milliseconds and needs to be converted into
+			 * a timeval structure.
+			 */
+			int timeout_us;
+			struct timeval tv;
+
+			timeout_us = atoi(sock_opt_value) * USEC_PER_MSEC;
+			tv.tv_sec = timeout_us / USEC_PER_SEC;
+			tv.tv_usec = timeout_us % USEC_PER_SEC;
+			err = setsockopt(socket_info->fd, sock_level, sock_opt_id, &tv,
+				sizeof(tv));
+			goto exit;
+
+		default:
+			break;
+		}
+		break; /* SOL_SOCKET */
+
+	default:
+		break;
+	}
+
+	/* All other values are integers. */
+	int sock_opt_value_int;
+
+	sock_opt_value_int = atoi(sock_opt_value);
+	err = setsockopt(socket_info->fd, sock_level, sock_opt_id, &sock_opt_value_int,
+				sizeof(sock_opt_value_int));
```

Comment:
I might be missing something with the indentations could this be in the default branch of switch-case and then we could get rid of `goto exit` in many places, if not all.

Reply:
Because there are nested switch-cases for level and option, the code would have to be in multiple default branches.

Reply:
True. Just let it be.